pkg/generator/gen.go

```diff
@@ -0,0 +1,340 @@
+
+		g.groupVersions = append(g.groupVersions, groups...)
+	}
+	return nil
+}
+
+// generate first genrates the wrapper for all the interfaces provided in the input.
+// Then for each type defined in the input, it recursively wraps the subsequent
+// interfaces to be kcp-aware.
+func (g *Generator) generate(ctx *genall.GenerationContext) error {
+	err := g.writeWrappedClientSet()
+	if err != nil {
+		return err
+	}
+	return g.generateSubInterfaces(ctx)
+}
+
+func (g *Generator) writeWrappedClientSet() error {
+	out := new(bytes.Buffer)
+	if err := wrtieHeader(out); err != nil {
+		return err
+	}
+
+	wrappedInf, err := internal.NewInterfaceWrapper(g.interfaceName, g.inputDir, g.groupVersions, out)
+	if err != nil {
+		return err
+	}
+
+	if err := wrappedInf.WriteContent(); err != nil {
+		return err
+	}
+
+	outBytes := out.Bytes()
+	formattedBytes, err := format.Source(outBytes)
+	if err != nil {
+		return err
+	} else {
+		outBytes = formattedBytes
+	}
+	return g.writeContent(outBytes, clientSetFilename)
+}
+
+// wrtieContents creates a new file under the path <outputDir>/generated with
+// the specified filename and write contents to it.
+func (g *Generator) writeContent(outBytes []byte, filename string) error {
+	path := filepath.Join(g.outputDir, generatedFolderName)
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		err = os.MkdirAll(path, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+
+	outputFile, err := os.Create(filepath.Join(path, filename))
+	if err != nil {
+		return err
+	}
+	defer outputFile.Close()
+
+	n, err := outputFile.Write(outBytes)
+	if err != nil {
+		return err
+	}
+	if n < len(outBytes) {
+		return err
+	}
+	return nil
+}
+
+func wrtieHeader(out io.Writer) error {
+	n, err := out.Write([]byte(internal.HeaderText))
+	if err != nil {
+		return err
+	}
+
+	if n < len([]byte(internal.HeaderText)) {
+		return errors.New("header text was not written properly.")
+	}
+	return nil
+}
+
+func (g *Generator) generateSubInterfaces(ctx *genall.GenerationContext) error {
+	for _, gv := range g.groupVersions {
+		// the group version flag from input is allowed to have only one version
+		// for now.
+		// TODO: modify this to accept a list instead
+		version := gv.Versions[0]
+		path := filepath.Join(g.inputDir, "pkg", "apis", gv.Group.String(), string(version.Version))
+
+		pkgs, err := loader.LoadRoots(path)
+		if err != nil {
+			return err
+		}
+
+		for _, root := range pkgs {
+			root.NeedTypesInfo()
+
+			// this is to accomodate multiple types defined in single group
+			byType := make(map[string][]byte)
+
+			outCommonContent := new(bytes.Buffer)
+			pkgmg, err := internal.NewPacakges(root, path, g.clientSetAPIPath, string(version.Version), gv.PackageName, outCommonContent)
+			if err != nil {
+				return err
+			}
+
+			if err := wrtieHeader(outCommonContent); err != nil {
+				return err
+			}
+			err = pkgmg.WriteContent()
+			if err != nil {
+				return err
+			}
+
+			var e error
```

Comment:
I'm not sure what the intent is here with `e`. I see the error returned from calling `markers.EachType` is assigned to `e`, but then inside the anonymous function closure, you're also assigning `err` to `e` (which I assume is lost when `EachType` returns).

Also, we typically have error variables end with Err, such as `eachTypeErr`, if you need to distinguish between 2 different `error` variables.

Reply:
The issue we are facing here is that [markers.EachType](https://pkg.go.dev/sigs.k8s.io/controller-tools/pkg/markers#EachType) is defined to accept a call back func which does not return an error. Is there a better way to propagate the error? One option is to append errors for each type in a map which has the key to be the type name and eventually return it in the end. But this may not be the best way to do that since we would end up returning only the error for first type we encounter. Can you please provide some inputs on how to handle it better?

Reply:
You definitely have the right idea! I just think it needs some minor tweaking. Here is one possibility:

```go
var errs []error
if err := markers.EachType(ctx.Collector, root, func(info *markers.TypeInfo) {
  if err := doSomething(); err != nil {
    errs = append(errs, err)
  }
  // ...
}; err != nil {
  errs = append(errs, err)
}

if len(errs) > 0 {
  return errors.NewAggregate(errs)
}
```

Reply:
There's also `root.AddError(err)` but I'm not sure without more digging how/where that's used.

Reply:
Looked into this a little bit and looks like controller-tools `root.AddError(err)` is actually making things easy. It has an option to consolidate errors from each type and [print](https://pkg.go.dev/sigs.k8s.io/controller-tools/pkg/loader#PrintErrors) in eventually. Used the same logic here.